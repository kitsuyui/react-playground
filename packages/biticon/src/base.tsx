import React from 'react'
import { COLORS } from './colors'
import { applySBox128bitBigintTo128bitBigint } from './bitShuffle'


export const UUIDIcon = (props: { uuid: string }): React.JSX.Element => {
  const uuid = props.uuid
  if (!isValidUUID(uuid)) {
    return <span>Invalid UUID</span>
  }
  const bits = uuidToBigInt(uuid)
  const newBits = applySBox128bitBigintTo128bitBigint(bits)
  return <BitIcon bits={newBits} />
}

const isValidUUID = (uuid: string): boolean => {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/.test(uuid)
}

const uuidToBigInt = (uuid: string): bigint => {
  // strip if it has dashes
  const uuid_ = uuid.replace(/-/g, '')
  return BigInt(`0x${uuid_}`)
}


export const BitIcon = (props: { bits: bigint }): React.JSX.Element => {
  const bits = props.bits
  if (!is128Bit(bits)) {
    return (<span>Invalid 128bit value</span>)
  }
  const items = from128bitTo8bitArray(bits)
  const width = 64
  const height = 64
  const numOfItems = items.length
  const numOfRows = Math.ceil(Math.sqrt(numOfItems))
  const numOfCols = Math.ceil(numOfItems / numOfRows)
  const itemWidth = width / numOfCols
  const itemHeight = height / numOfRows
  const text = items.reduce((acc, item) => acc + item.toString(16), '')

  return <svg width={width} height={height} viewBox={`0 0 ${width} ${height}`} xmlns="http://www.w3.org/2000/svg" aria-label={`BitIcon: ${text}`} role="img">
    {Array.from(items).map((item: number, index: number) => {
      const indexOfX = index % numOfCols // 0-7
      const indexOfY = Math.floor(index / numOfCols) // 0-7
      const x = indexOfX * itemWidth
      const y = indexOfY * itemHeight
      const color = valueToColor(item)
      // biome-ignore lint/suspicious/noArrayIndexKey: SVG elements require a unique key
      return <rect key={index} x={x} y={y} width={itemWidth} height={itemHeight} fill={color} />
    })}
  </svg>
}

/**
 * Convert value to color
 * @param value 0-7
 * @returns color string
 */
const valueToColor = (value: number): string => {
  return COLORS[value]
}

const is128Bit = (value: bigint): boolean => {
  return value >= BigInt(0) && value <= BigInt('0xffffffffffffffffffffffffffffffff')
}

/**
 * Convert 128bit value to int8 array
 * @param value 128bit value
 * @returns int8 array of 64 elements (0-7)
 */
export const from128bitTo8bitArray = (value: bigint): Uint8Array => {
  let value_ = value
  const bytes = new Uint8Array(64)
  for (let i = 0; i < 64; i++) {
    // take 2 bits from the value
    const bit = Number(value_ & BigInt(0b11)) // 0-3
    // shift 2 bits
    value_ >>= BigInt(2)
    bytes[i] = bit
  }
  return bytes
}
