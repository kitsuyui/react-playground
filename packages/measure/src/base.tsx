import React, { useMemo } from 'react'
import { createCrossLinesBySize } from './calc'


interface TextStyle {
  fontFamily: string
}

interface StrokeStyle {
  color: string
  width: number
  dasharray: number
}

export const DEFAULT_MEASURE_STYLE: MeasureStyle = {
  stroke: {
    color: 'black',
    width: 1,
    dasharray: 2,
  },
  text: {
    fontFamily: 'monospace',
  },
} as const

export interface MeasureStyle {
  stroke: StrokeStyle
  text: TextStyle
}

/**
 * MeasureBase component
 *
 * This component is a base component for Measure component.
 * It draws a rectangle, cross lines, and text.
 * It does not have any state and side effect. Pure.
 *
 * @param props.width {number}
 * @param props.height {number}
 * @param props.styles {MeasureStyle}
 * @returns
 */
export const MeasureBase = (props: { width: number, height: number, styles: MeasureStyle }): React.JSX.Element => {
  const { width, height, styles } = props
  const text = `${width}x${height}`
  const textSize = text.length
  const fontSize = Math.min(width, height) / textSize
  const [centerX, centerY] = [width / 2, height / 2]
  const stroke = styles.stroke
  const lines = useMemo(() => createCrossLinesBySize({ width, height }), [width, height])
  return (
    <svg width={width} height={height}>
      <title>{text}</title>
      <rect
        x={0}
        y={0}
        width={width}
        height={height}
        fill="none"
        stroke={stroke.color}
        strokeWidth={stroke.width}
        strokeDasharray={stroke.dasharray}
      />
      {
        lines.map((line, index) => (
          <line
            // biome-ignore lint/suspicious/noArrayIndexKey: lines are always 4 items, so no need to add key
            key={index}
            x1={line.start.x}
            y1={line.start.y}
            x2={line.end.x}
            y2={line.end.y}
            stroke={stroke.color}
            strokeWidth={stroke.width}
            strokeDasharray={stroke.dasharray}
          />
        ))
      }
      <text
        x={centerX}
        y={centerY}
        textAnchor="middle"
        dominantBaseline="middle"
        fontSize={fontSize}
        fontFamily={styles.text.fontFamily}
      >
        {text}
      </text>
    </svg>
  )
}
