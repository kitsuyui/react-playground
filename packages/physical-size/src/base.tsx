import React from 'react'

import { createContext, useEffect, useRef, useState, forwardRef } from 'react'

/**
 * I want to get the physical size of the device.
 * But this package is meaningless.Because PPI is constant value for each device.
 * In CSS definition, 1in is always 96px.So PPI = sqrt(96 * 96) = 96.
 * https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units
 */

export interface PhysicalSize {
  ppi: number
}

const DEFAULT_PPI = 96

/**
 * PhysicalSizeContext is a context that provides the physical size of the device.
 */
export const PhysicalSizeContext = createContext<PhysicalSize>({
  ppi: DEFAULT_PPI,
})

/**
 * PhysicalSizeProvider is a provider that provides the physical size of the device.
 * @param props children
 * @returns PhysicalSizeProvider
 */
export const PhysicalSizeProvider = (props: { children: React.ReactNode }) => {
  const ref = useRef<HTMLDivElement>(null)
  const [ppi, setPpi] = useState(DEFAULT_PPI)
  useEffect(() => {
    if (!ref?.current) return
    const ppi = calcPPIFromHiddenComponent(ref.current)
    setPpi(ppi)
  })
  return (
    <>
      <HiddenComponent ref={ref} />
      <PhysicalSizeContext.Provider value={{ ppi }}>
        {props.children}
      </PhysicalSizeContext.Provider>
    </>
  )
}

export const calcPPI = (values: { width: number; height: number }): number => {
  const { width, height } = values
  return Math.sqrt(width * height)
}

/**
 * HiddenComponent is a hidden component that is used to calculate PPI.
 * @param props ref
 * @returns HiddenComponent
 */
const HiddenComponent = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <div
      ref={ref}
      style={{
        width: '1in',
        height: '1in',
        position: 'absolute',
        top: '-100%',
        left: '-100%',
        opacity: 0,
        overflow: 'hidden',
      }}
    />
  )
})

/**
 * calculate PPI from hidden component
 * @param element hidden component
 * @returns PPI (Pixel Per Inch)
 */
const calcPPIFromHiddenComponent = (element: HTMLDivElement): number => {
  const { offsetWidth, offsetHeight } = element
  return calcPPI({ width: offsetWidth, height: offsetHeight })
}
