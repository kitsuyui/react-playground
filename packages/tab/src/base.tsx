/**
 * Abstract Tab UI component
 *
 * This component is a base for creating tab UI components.
 * This component doesn't have any styles, positions, states, or any other details.
 * This component only provides an ideal structure for tab UI and its ReactContext.
 *
 * So, you can customize the tab UI as you want.
 * Or you can compose this component with other components and CSS modules.
 */
import React from 'react'
import { createContext } from 'react'

export type TabItemId = string
export type TabTitle = React.ReactNode | string

export interface TabItem {
  id: TabItemId
  title: TabTitle
  content: React.ReactNode
}

export interface TabUIComponents {
  TabBar: React.ComponentType<{ children: React.ReactNode }>
  TabButton: React.ComponentType
  ContentBox: React.ComponentType<{ children: React.ReactNode }>
}

export interface TabControlProps {
  items: TabItem[]
  selectedTabId: TabItemId
  onSelect: (id: TabItemId) => void
}

export interface TabButtonContextValue {
  id: TabItemId
  onSelect: (id: TabItemId) => void
  selected: boolean
  title: TabTitle
}

const DEFAULT_TAB_BUTTON_CONTEXT: TabButtonContextValue = {
  id: '',
  onSelect: () => { },
  selected: false,
  title: '',
} as const

export const TabButtonContext = createContext<TabButtonContextValue>(
  DEFAULT_TAB_BUTTON_CONTEXT,
)

export const TabUIBase = (props: TabControlProps & TabUIComponents) => {
  const { items, selectedTabId, onSelect, TabBar, TabButton, ContentBox } =
    props
  const activeItem = items.find((tab) => tab.id === selectedTabId)
  return (
    <>
      <TabBar>
        {items.map((item) => (
          <TabButtonContext.Provider
            key={item.id}
            value={{
              id: item.id,
              title: item.title,
              selected: item.id === selectedTabId,
              onSelect,
            }}
          >
            <TabButton />
          </TabButtonContext.Provider>
        ))}
      </TabBar>
      <ContentBox>{activeItem?.content}</ContentBox>
    </>
  )
}
