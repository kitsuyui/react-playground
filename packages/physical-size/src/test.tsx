import { render } from '@testing-library/react'
import React from 'react'

import { PhysicalSizeProvider, PhysicalSizeContext, calcPPI } from './base'

describe('PhysicalSizeProvider', () => {
  it('should render children', () => {
    const { container } = render(
      <PhysicalSizeProvider>
        <div>children</div>
      </PhysicalSizeProvider>,
    )
    expect(container).toMatchSnapshot()
  })
})

describe('PhysicalSizeContext', () => {
  it('should render children', () => {
    const { container } = render(
      <PhysicalSizeProvider>
        <PhysicalSizeContext.Consumer>
          {(value) => <div>{value.ppi}</div>}
        </PhysicalSizeContext.Consumer>
      </PhysicalSizeProvider>,
    )
    expect(container).toMatchSnapshot()
  })
})

describe('calcPPI', () => {
  it('should return the correct PPI', () => {
    expect(calcPPI({ width: 96, height: 96 })).toBe(96)
    expect(calcPPI({ width: 192, height: 192 })).toBe(192)
    // weird test case
    expect(calcPPI({ width: 96, height: 192 })).toBe(135.7645019878171)
  })
})
