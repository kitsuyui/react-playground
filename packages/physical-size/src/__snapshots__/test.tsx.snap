// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhysicalSizeContext should render children 1`] = `
<div>
  <div
    style="width: 1in; height: 1in; position: absolute; top: -100%; left: -100%; opacity: 0; overflow: hidden;"
  />
  <div>
    0
  </div>
</div>
`;

exports[`PhysicalSizeProvider should render children 1`] = `
<div>
  <div
    style="width: 1in; height: 1in; position: absolute; top: -100%; left: -100%; opacity: 0; overflow: hidden;"
  />
  <div>
    children
  </div>
</div>
`;
